drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,2019 @@
+			msg.type = NUMAKER_USBD_MSG_TYPE_CB_STATE;
+			msg.cb_device.status_code = USB_DC_RESET;
+			numaker_usbd_send_msg(dev, &msg);
+
+			LOG_DBG("USB reset");
+		}
+		if (usbd_bus_state & USBD_STATE_SUSPEND) {
+			/* Enable USB but disable PHY */
+			config->base->ATTR &= ~USBD_PHY_EN;
+
+			/* Message for bottom-half processing */
+			msg.type = NUMAKER_USBD_MSG_TYPE_CB_STATE;
+			msg.cb_device.status_code = USB_DC_SUSPEND;
+			numaker_usbd_send_msg(dev, &msg);
+
+			LOG_DBG("USB suspend");
+		}
+		if (usbd_bus_state & USBD_STATE_RESUME) {
+			/* Enable back USB/PHY */
+			config->base->ATTR |= USBD_ATTR_USBEN_Msk | USBD_ATTR_PHYEN_Msk;
+
+			/* Message for bottom-half processing */
+			msg.type = NUMAKER_USBD_MSG_TYPE_CB_STATE;
+			msg.cb_device.status_code = USB_DC_RESUME;
+			numaker_usbd_send_msg(dev, &msg);
+
+			LOG_DBG("USB resume");
+		}
+	}
+
+	/* USB SOF */
+	if (usbd_intsts & USBD_INTSTS_SOFIF_Msk) {
+		/* Clear event flag */
+		config->base->INTSTS = USBD_INTSTS_SOFIF_Msk;
+
+		/* Message for bottom-half processing */
+		msg.type = NUMAKER_USBD_MSG_TYPE_CB_STATE;
+		msg.cb_device.status_code = USB_DC_SOF;
+		numaker_usbd_send_msg(dev, &msg);
+	}
+
+	/* USB Setup/EP */
+	if (usbd_intsts & USBD_INTSTS_USB) {
+		uint32_t epintsts;
+
+		/* Setup event */
+		if (usbd_intsts & USBD_INTSTS_SETUP) {
+			/* Clear event flag */
+			config->base->INTSTS = USBD_INTSTS_SETUP;
+
+			/* Clear the data IN/OUT ready flag of control endpoints */
+			NUMAKER_USBD_EP_BASE(config->base, EP0)->CFGP |= USBD_CFGP_CLRRDY_Msk;
+			NUMAKER_USBD_EP_BASE(config->base, EP1)->CFGP |= USBD_CFGP_CLRRDY_Msk;
+
+			/* By USB spec, following transactions, regardless of Data/Status stage,
+			 * will always be DATA1
+			 */
+			NUMAKER_USBD_EP_BASE(config->base, EP0)->CFG |= USBD_CFG_DSQSYNC_Msk;
+			NUMAKER_USBD_EP_BASE(config->base, EP1)->CFG |= USBD_CFG_DSQSYNC_Msk;
+
+			/* Message for bottom-half processing */
+			/* NOTE: In Zephyr USB device stack, Setup packet is passed via
+			 * CTRL OUT EP
+			 */
+			msg.type = NUMAKER_USBD_MSG_TYPE_CB_EP;
+			msg.cb_ep.ep = USB_EP_GET_ADDR(0, USB_EP_DIR_OUT);
+			msg.cb_ep.status_code = USB_DC_EP_SETUP;
+			numaker_usbd_send_msg(dev, &msg);
+		}
+
+		/* EP events */
+		epintsts = config->base->EPINTSTS;
+
+		config->base->EPINTSTS = epintsts;
+
+		while (epintsts) {
+			uint32_t hw_ep_idx = u32_count_trailing_zeros(epintsts);
+
+			/* We don't enable INNAKEN interrupt, so as long as EP event occurs,
+			 * we can just regard one data transaction has completed (ACK for
+			 * CTRL/BULK/INT or no-ACK for Iso), that is, no need to check EPSTS0,
+			 * EPSTS1, etc.
+			 */
+
+			/* EP direction, number, and address */
+			uint8_t ep_dir = ((config->base->EP[hw_ep_idx].CFG & USBD_CFG_STATE_Msk) ==
+					  USBD_CFG_EPMODE_IN)
+						 ? USB_EP_DIR_IN
+						 : USB_EP_DIR_OUT;
+			uint8_t ep_idx = (config->base->EP[hw_ep_idx].CFG & USBD_CFG_EPNUM_Msk) >>
+					 USBD_CFG_EPNUM_Pos;
+			uint8_t ep = USB_EP_GET_ADDR(ep_idx, ep_dir);
```

Comment:
```c
			uint32_t hw_ep_idx = u32_count_trailing_zeros(epintsts);
			uint8_t ep;

			ep = (base->EP[hw_ep_idx].CFG & USBD_CFG_EPNUM_Msk) >>
			     USBD_CFG_EPNUM_Pos;
			if ((base->EP[hw_ep_idx].CFG & USBD_CFG_STATE_Msk) ==
			    USBD_CFG_EPMODE_IN) {
				ep |= USB_EP_DIR_IN;
			} 
```

Ideally, this entire block can be moved to its own inline function.

Reply:
used inline functions and made suggested changes